contracts/utils/RetireToucanCarbon.sol

```diff
@@ -0,0 +1,1625 @@
+        if (_selection == 4) {
+            stakingHelper = _newAddress; // 4; Set new stakingHelper address
+        }
+
+        return true;
+    }
+
+    /**
+        @notice Update the router for an existing pool
+        @param _amount New fee amount, in .1% increments. 10 = 1%
+        @return bool
+     */
+    function setFeeAmount(uint256 _amount) external onlyOwner returns (bool) {
+        uint256 oldFee = feeAmount;
+        feeAmount = _amount;
+
+        emit FeeUpdated(oldFee, feeAmount);
+        return true;
+    }
+
+    /**
+        @notice Update the router for an existing pool
+        @param _toucanPool Pool being updated
+        @param _router New router address
+        @return bool
+     */
+    function setPoolRouter(address _toucanPool, address _router)
+        external
+        onlyOwner
+        returns (bool)
+    {
+        address oldRouter = poolRouter[_toucanPool];
+        poolRouter[_toucanPool] = _router;
+        emit PoolRouterChanged(_toucanPool, oldRouter, poolRouter[_toucanPool]);
+        return true;
+    }
+
+    /**
+        @notice Add a new carbon pool to retire with helper contract
+        @param _toucanPool Pool being added
+        @param _router UniswapV2 router to route trades through for non-pool retirements
+        @param _bondDepository Bond depository address
+        @return bool
+     */
+    function addPool(
+        address _toucanPool,
+        address _router,
+        address _bondDepository
+    ) external onlyOwner returns (bool) {
+        bool result;
+
+        require(!listContains(toucanPools, _toucanPool), "Pool already added");
+        require(_toucanPool != address(0), "Pool cannot be zero address");
+
+        toucanPools.push(_toucanPool);
+        isToucanPool[_toucanPool] = true;
+        poolRouter[_toucanPool] = _router;
+        bondDepository[_toucanPool] = _bondDepository;
+
+        emit PoolAdded(_toucanPool, _router, _bondDepository);
+        return true;
+    }
+
+    /**
+        @notice Update the bond depository for a carbon pool
+        @param _toucanPool Pool being updated
+        @param _bondDepository New depository address
+        @return bool
+     */
+    function updateBondDepository(address _toucanPool, address _bondDepository)
+        external
+        onlyOwner
+        returns (bool)
+    {
+        require(isToucanPool[_toucanPool], "Not a Toucan Carbon Pool");
+
+        address oldBond = bondDepository[_toucanPool];
+        bondDepository[_toucanPool] = _bondDepository;
+        emit PoolBondChanged(_toucanPool, oldBond, bondDepository[_toucanPool]);
+        return true;
+    }
+
+    /**
+        @notice checks array to ensure against duplicate
+        @param _list address[]
+        @param _token address
+        @return bool
+     */
+    function listContains(address[] storage _list, address _token)
+        internal
+        view
+        returns (bool)
+    {
+        for (uint256 i = 0; i < _list.length; i++) {
```

Comment:
Another storage read optimization:
```js
uint256 listLen = _list.length;
for (uint256 i = 0; i < listLen; i++) {
```
https://gist.github.com/hrkrshnn/ee8fabd532058307229d65dcd5836ddc#caching-the-length-in-for-loops